File: User/ugcp_svc/Content/Profile/CProfile.cpp
#include "../framework.h"
#include <setupapi.h>
#include <initguid.h>
#include "CProfile.h"
#include "../EventQueue/CEventQueue.h"
#include "../ProcessOutput/CProcessOutput.h"
#include "../X52/USBX52Write.h"
#include "../X52/MFDMenu.h"
#include "../ProcessOutput/CVirtualHID.h"


CProfile::CProfile()
{
	hMutexProgram = CreateMutex(NULL, FALSE, NULL);
	hMutexCalibration = CreateMutex(NULL, FALSE, NULL);
	hMutexStatus = CreateMutex(NULL, FALSE, NULL);
	status.Mode = 0;
	status.SubMode = 0;
	profile.MouseTick = 10;
	rawMode = FALSE;
	calibrationMode = FALSE;
}

CProfile::~CProfile()
{
	WaitForSingleObject(hMutexCalibration, INFINITE);
	calibration.Limits.clear();
	calibration.Jitters.clear();
	CloseHandle(hMutexCalibration);
	ClearProfile();
	CloseHandle(hMutexProgram);
	CloseHandle(hMutexStatus);
}

void CProfile::ClearProfile()
{
	WaitForSingleObject(hMutexProgram, INFINITE);
	{
		if (CEventQueue::Get() != nullptr)
		{
			CEventQueue::Get()->Clear();
		}
		if (CProcessOutput::Get() != nullptr)
		{
			CProcessOutput::Get()->ClearEvents();
		}

		if (profile.Actions != nullptr)
		{
			while (!profile.Actions->empty())
			{
				delete profile.Actions->front();
				profile.Actions->pop_front();
			}
			delete profile.Actions;
			profile.Actions = nullptr;
		}

		profile.Clear();

		LockStatus();
		{
			status.Clear();
		}
		UnlockStatus();
		newProfileIn = true;
		newProfileOut = true;
	}
	ReleaseMutex(hMutexProgram);
}

bool CProfile::HF_IoWriteMap(std::span<const std::uint8_t> SystemBuffer)
{
	if (!resetComanmds)
	{
		return false;
	}
	resetComanmds = false;
	if (SystemBuffer.size() < (17 + 1 + 2)) //txt, mouse, button sz, axes sz
	{
		ClearProfile();
		return false;
	}
	else
	{
		CX52Write::Get().Set_Text(SystemBuffer.subspan(0, 17));
		static constexpr std::uint8_t cmd2[] = { 2, 0 };
		CX52Write::Get().Set_Text(cmd2);
		static constexpr std::uint8_t cmd3[] = { 3, 0 };
		CX52Write::Get().Set_Text(cmd3);
	}

	std::unordered_map<UINT32, char> ids;

	WaitForSingleObject(hMutexProgram, INFINITE);
	{
		const std::uint8_t* ptr = SystemBuffer.data() + 17;
		RtlCopyMemory(&profile.MouseTick, ptr++ , 1);

		BYTE count = *ptr++;
		for (BYTE j = 0; j < *(SystemBuffer.data() + 17 + 1); j++)
		{
			UINT32 joyId = *((UINT32*)ptr);
			ids.insert({ joyId, 0 });
			ptr += sizeof(UINT32);			
			for (BYTE nModes = *ptr++; nModes > 0; nModes--)
			{
				BYTE mode = *ptr++;
				for (BYTE nButtons = *ptr++; nButtons > 0; nButtons--)
				{
					BYTE buttonId = *ptr++;
					std::vector<UINT16>* pActions = profile.ButtonsMap.NewButton(joyId, mode, buttonId, *ptr++);
					status.Buttons.NewStatus(joyId, mode, buttonId);

					for (BYTE nActions = *ptr++; nActions > 0; nActions--)
					{
						pActions->push_back(*((UINT16*)ptr));
						ptr += sizeof(UINT16);
					}
				}
			}
		}

		count = *ptr++;
		for (BYTE j = 0; j < count; j++)
		{
			UINT32 joyId = *((UINT32*)ptr);
			ids.insert({ joyId, 0 });
			ptr += sizeof(UINT32);
			for (BYTE nModes = *ptr++; nModes > 0; nModes--)
			{
				BYTE mode = *ptr++;
				for (BYTE nButtons = *ptr++; nButtons > 0; nButtons--)
				{
					BYTE buttonId = *ptr++;
					std::vector<UINT16>* pActions = profile.HatsMap.NewButton(joyId, mode, buttonId, *ptr++);
					status.Hats.NewStatus(joyId, mode, buttonId);

					for (BYTE nActions = *ptr++; nActions > 0; nActions--)
					{
						pActions->push_back(*((UINT16*)ptr));
						ptr += sizeof(UINT16);
					}
				}
			}
		}

		count = *ptr++;
		for (BYTE j = 0; j < count; j++)
		{
			UINT32 joyId = *((UINT32*)ptr);
			ids.insert({joyId, 0});
			ptr += sizeof(UINT32);
			for (BYTE nModes = *ptr++; nModes > 0; nModes--)
			{
				BYTE mode = *ptr++;
				for (BYTE nAxes = *ptr++; nAxes > 0; nAxes--)
				{
					BYTE axisKey = *ptr++;
					PROGRAMMING::AXISMODEL* axis = profile.AxesMap.NewAxis(joyId, mode, axisKey);
					status.Axes.NewStatus(joyId, mode, axisKey);

					axis->MouseSensibility = *ptr++;
					axis->VJoyOutput = *ptr++;
					axis->Type = *ptr++;
					axis->OutputAxis = *ptr++;
					axis->SoftDeadZone = *ptr++;
					memcpy(&axis->SensibilityX, ptr, 20 * sizeof(double));
					ptr += 20 * sizeof(double);
					memcpy(&axis->SensibilityY, ptr, 20 * sizeof(double));
					ptr += 20 * sizeof(double);
					memcpy(&axis->SensibilityS, ptr, 20 * sizeof(double));
					ptr += 20 * sizeof(double);
					axis->IsSlider = *ptr++;
					//RtlCopyMemory(&axis->Inertia, ptr,  sizeof(double));
					//ptr += sizeof(double);
					memcpy(&axis->DampingK, ptr, sizeof(double));
					ptr += sizeof(double);
					for (BYTE nBands = *ptr++; nBands > 0; nBands--)
					{
						axis->Bands.push_back(*ptr++);
					}
					for (BYTE nActions = *ptr++; nActions > 0; nActions--)
					{
						axis->Actions.push_back(*((UINT16*)ptr));
						ptr += sizeof(UINT16);
					}
					axis->ToughnessInc = *ptr++;
					axis->ToughnessDec = *ptr++;
				}
			}
		}
	}
	ReleaseMutex(hMutexProgram);

	UINT32* newDevices = new UINT32[ids.size()];
	UCHAR i = 0;
	while (ids.size() > 0)
	{
#pragma warning(disable:6386)
		newDevices[i++] = ids.begin()->first;
#pragma warning(default:6386)
		ids.erase(ids.begin()->first);
	}
	if (RefreshHidInputDevice != nullptr)
	{
		RefreshHidInputDevice(hidInput, newDevices, i);
	}
	delete[] newDevices;

	return true;
}

bool CProfile::HF_IoWriteCommands(std::span<const std::uint8_t> SystemBuffer)
{
	const std::uint8_t* bufIn;
	size_t sizePredicted = 0;

	ClearProfile();

	if (SystemBuffer.size() != 0)
	{
		//Check OK
		bufIn = SystemBuffer.data();
		while (sizePredicted < SystemBuffer.size())
		{
			sizePredicted += (static_cast<size_t>(*bufIn) * 4) + 1;
			bufIn += (static_cast<size_t>(*bufIn) * 4) + 1;
		}
		if (sizePredicted != SystemBuffer.size())
		{
			return false;
		}
	}

	resetComanmds = true;

	CMFDMenu::Get().SetHourActivated(true);
	CMFDMenu::Get().SetDateActivated(true);

	if (SystemBuffer.size() == 0)
	{
		return true;
	}

	WaitForSingleObject(hMutexProgram, INFINITE);
	{
		profile.Actions = new std::deque<CEventPacket*>();
		bufIn = SystemBuffer.data();
		sizePredicted = 0;
		while (sizePredicted < SystemBuffer.size())
		{
			CEventPacket* commandQueue = new CEventPacket();
			UCHAR actionSize = *bufIn;
			UCHAR i = 0;

			bufIn++;
			sizePredicted++;

			for (i = 0; i < actionSize; i++)
			{
				PEV_COMMAND mem = new EV_COMMAND;
				RtlZeroMemory(mem, sizeof(EV_COMMAND));
				mem->Type = *bufIn;
				mem->Basic.Data1 = *(bufIn + 1);
				mem->Basic.Data2 = *(bufIn + 2);
				mem->Basic.Extra = *(bufIn + 3) & 0xf;
				mem->Basic.OutputJoy = *(bufIn + 3) >> 4;
				if ((((PEV_COMMAND)bufIn)->Type == CommandType::X52MfdHour) || (((PEV_COMMAND)bufIn)->Type == CommandType::X52MfdHour24))
				{
					CMFDMenu::Get().SetHourActivated(false);
				}
				else if (((PEV_COMMAND)bufIn)->Type == CommandType::MfdDate)
				{
					CMFDMenu::Get().SetDateActivated(false);
				}
				bufIn += 4;
				sizePredicted += 4;
				commandQueue->AddCommand(mem);
			}
			profile.Actions->push_back(commandQueue);
		}
	}
	ReleaseMutex(hMutexProgram);

	return true;
}

void CProfile::WriteAntivibration(std::span<const std::uint8_t> data)
{
	WaitForSingleObject(hMutexCalibration, INFINITE);

	const std::uint8_t* ptr = data.data();
	ptr += sizeof(UINT32);
	for (UINT32 njoys = 0; njoys < *reinterpret_cast<const UINT32*>(data.data()); njoys++)
	{
		std::unordered_map<UINT32, std::vector<CALIBRATION::ST_JITTER>>::iterator pOld = calibration.Jitters.find(*reinterpret_cast<const UINT32*>(ptr));
		if (pOld != calibration.Jitters.end())
		{
			calibration.Jitters.erase(*reinterpret_cast<const UINT32*>(ptr));
		}
		calibration.New.insert({ *reinterpret_cast<const UINT32*>(ptr), true }).first->second = true;
		std::vector<CALIBRATION::ST_JITTER>* vector = &calibration.Jitters.insert({ *reinterpret_cast<const UINT32*>(ptr), std::vector<CALIBRATION::ST_JITTER>()}).first->second;
		ptr += sizeof(UINT32);
		BYTE naxes = *ptr++;
		for (BYTE njitters = 0; njitters < naxes; njitters++)
		{
			vector->push_back(CALIBRATION::ST_JITTER());
			RtlCopyMemory(&vector->at(njitters), ptr, sizeof(CALIBRATION::ST_JITTER));
			ptr += sizeof(CALIBRATION::ST_JITTER);
		}
	}
	ReleaseMutex(hMutexCalibration);
}

void CProfile::WriteCalibration(std::span<const std::uint8_t> data)
{
	WaitForSingleObject(hMutexCalibration, INFINITE);

	const std::uint8_t* ptr = data.data();
	ptr += sizeof(UINT32);
	for (UINT32 njoys = 0; njoys < *reinterpret_cast<const UINT32*>(data.data()); njoys++)
	{
		std::unordered_map<UINT32, std::vector<CALIBRATION::ST_LIMITS>>::iterator pOld = calibration.Limits.find(*reinterpret_cast<const UINT32*>(ptr));
		if (pOld != calibration.Limits.end())
		{
			calibration.Limits.erase(*reinterpret_cast<const UINT32*>(ptr));
		}
		calibration.New.insert({ *reinterpret_cast<const UINT32*>(ptr), true }).first->second = true;
		std::vector<CALIBRATION::ST_LIMITS>* vector = &calibration.Limits.insert({ *reinterpret_cast<const UINT32*>(ptr), std::vector<CALIBRATION::ST_LIMITS>()}).first->second;
		ptr += sizeof(UINT32);
		BYTE naxes = *ptr++;
		for (BYTE nlimits = 0; nlimits < naxes; nlimits++)
		{
			vector->push_back(CALIBRATION::ST_LIMITS());
			RtlCopyMemory(&vector->at(nlimits), ptr, sizeof(CALIBRATION::ST_LIMITS));
			vector->back().Range = vector->back().Right;
			ptr += sizeof(CALIBRATION::ST_LIMITS);
		}
	}
	ReleaseMutex(hMutexCalibration);
}

File: User/ugcp_svc/Profile/CProfile.cs
using System;
using System.Collections.Generic;

namespace ugcp_svc.Profile
{
    sealed class CProfile : IDisposable
    {
        private readonly CCalibration calibration = new();
        private CProgramming profile = new();
        private CStatus status = new();
        private byte _rawMode = 0;
        //private byte _calibrationMode = 0;
        private Action<HashSet<uint>>? RefreshHidInputDevice = null;
        private Action<bool>? PauseWinUSB = null;

        void IDisposable.Dispose()
        {
            ((IDisposable)calibration).Dispose();
            ClearProfile();
        }

        private void ClearProfile()
        {
            CProgramming closeProfile = new();
            System.Threading.Interlocked.Exchange(ref profile, closeProfile);
            CStatus closeStatus = new();
            System.Threading.Interlocked.Exchange(ref status, closeStatus);
            ProcessOutput.CProcessOutput.Get()?.ClearEvents();
        }

        //public bool CalibrationMode
        //{
        //    get => System.Threading.Interlocked.CompareExchange(ref _calibrationMode, 0, 0) == 1;
        //    set
        //    {
        //        System.Threading.Interlocked.Exchange(ref _calibrationMode, (byte)(value ? 1 : 0));
        //        PauseWinUSB?.Invoke(value);
        //    }
        //}

        public bool RawMode
        {
            get => System.Threading.Interlocked.CompareExchange(ref _rawMode, 0, 0) == 1;
            set
            {
                System.Threading.Interlocked.Exchange(ref _rawMode, (byte)(value ? 1 : 0));
                PauseWinUSB?.Invoke(value);
            }
        }

        public CProgramming GetProfile() { return profile; }

        public CCalibration GetCalibration() { return calibration; }

        public CStatus GetStatus() { return status; }

        public void SetRefreshDevicesCallback(Action<HashSet<uint>> FnRefreshHidInputDevice)
        {
            RefreshHidInputDevice = FnRefreshHidInputDevice;
        }

        public void SetPauseWinUSBCallback(Action<bool> FnPauseWinUSB)
        {
            PauseWinUSB = FnPauseWinUSB;
        }

        public bool WriteProfile(ReadOnlySpan<char> SystemBuffer)
        {
            ClearProfile();
            CProgramming newProfile = new();

            if (SystemBuffer.Length < (17))
            {
                return false;
            }
            else
            {
                byte[] txtBuffer = new byte[17];
                txtBuffer[0] = (byte)SystemBuffer[0];
                string pfName = SystemBuffer.Slice(1, 16).ToString().TrimEnd('\0');
                pfName = pfName.Replace('ñ', 'ø').Replace('á', 'Ó').Replace('í', 'ß').Replace('ó', 'Ô').Replace('ú', 'Ò').Replace('Ñ', '£').Replace('ª', 'Ø').Replace('º', '×').Replace('¿', 'ƒ').Replace('¡', 'Ú').Replace('Á', 'A').Replace('É', 'E').Replace('Í', 'I').Replace('Ó', 'O').Replace('Ú', 'U');
                byte[] text = System.Text.Encoding.Convert(System.Text.Encoding.Unicode, System.Text.Encoding.GetEncoding(20127), System.Text.Encoding.Unicode.GetBytes(pfName));
                for (byte i = 0; i < 16; i++)
                {
                    if (text.Length >= (i + 1))
                        txtBuffer[i + 1] = text[i];
                    else
                        txtBuffer[i + 1] = 0;
                }

                txtBuffer[0] = 1;
                X52.CX52Write.Get()?.Set_Text(txtBuffer);
                ReadOnlySpan<byte> cmd2 = [2, 0];
                X52.CX52Write.Get()?.Set_Text(cmd2);
                ReadOnlySpan<byte> cmd3 = [3, 0];
                X52.CX52Write.Get()?.Set_Text(cmd3);
            }
            var jsonObject = System.Text.Json.JsonSerializer.Deserialize<Shared.ProfileModel>(SystemBuffer[17..]);
            if (jsonObject == null)
            {
                return false;
            }

            HF_IoWriteCommands(jsonObject.Macros, newProfile);
            HF_IoWriteMap(jsonObject, newProfile);
            System.Threading.Interlocked.Exchange(ref profile, newProfile);
            status.Reset(jsonObject);

            HashSet<uint> newDevices = [];
            foreach (uint d in jsonObject.ButtonsMap.Keys) { newDevices.Add(d); }
            foreach (uint d in jsonObject.HatsMap.Keys) { newDevices.Add(d); }
            foreach (uint d in jsonObject.AxesMap.Keys) { newDevices.Add(d); }

            RefreshHidInputDevice?.Invoke(newDevices);
            return true;
        }

        private static void HF_IoWriteMap(Shared.ProfileModel jsonObject, CProgramming newProfile)
        {
            newProfile.MouseTick = jsonObject.MouseTick;
            newProfile.ButtonsMap.LoadProfile(jsonObject.ButtonsMap);
            newProfile.HatsMap.LoadProfile(jsonObject.HatsMap);
            newProfile.AxesMap.LoadProfile(jsonObject.AxesMap);
        }

        private static void HF_IoWriteCommands(List<Shared.ProfileModel.MacroModel> jsonObject, CProgramming newProfile)
        {
            X52.CMFDMenu.Get()?.SetHourActivated(true);
            X52.CMFDMenu.Get()?.SetDateActivated(true);

            for (int m = 0; m < jsonObject.Count; m++)
            {
                var macro = jsonObject[m];
                EventQueue.EV_COMMAND[] cmds = new EventQueue.EV_COMMAND[macro.Commands.Count];

                for (int i = 0; i < cmds.LongLength; i++)
                {
                    ref var mem = ref cmds[i];
                    mem.Type = (byte)(macro.Commands[i] & 0xff);
                    mem.Data.Basic.Data1 = (byte)((macro.Commands[i] >> 8) & 0xff);
                    mem.Data.Basic.Data2 = (byte)((macro.Commands[i] >> 16) & 0xff);

                    byte extra = (byte)((macro.Commands[i] >> 24) & 0xff);
                    mem.Data.Basic.Extra = (byte)(extra & 0xf);
                    mem.Data.Basic.OutputJoy = (byte)(extra >> 4);
                    if ((mem.Type == EventQueue.CommandType.X52MfdHour) || (mem.Type == EventQueue.CommandType.X52MfdHour24))
                    {
                        X52.CMFDMenu.Get()?.SetHourActivated(false);
                    }
                    else if (mem.Type == EventQueue.CommandType.MfdDate)
                    {
                        X52.CMFDMenu.Get()?.SetDateActivated(false);
                    }
                }
                newProfile.WriteActions().Add(cmds);
            }
        }

        public void WriteAntivibration(ReadOnlySpan<char> ptr)
        {
            calibration.BeginCalibrationRead();
            {
                var jsonObject = System.Text.Json.JsonSerializer.Deserialize<List<Shared.Calibration.LowLevel.JoyJitters>>(ptr);
                if (jsonObject != null)
                {
                    calibration.Jitters = jsonObject;
                }
            }
            calibration.EndCalibrationRead();
        }

        public void WriteCalibration(ReadOnlySpan<char> ptr)
        {
            calibration.BeginCalibrationRead();
            {
                var jsonObject = System.Text.Json.JsonSerializer.Deserialize<List<Shared.Calibration.LowLevel.JoyLimits>>(ptr);
                if (jsonObject != null)
                {
                    calibration.Limits = jsonObject;
                }
            }
            calibration.EndCalibrationRead();
        }
    }
}
